File: leetcode/Q543_DiameterofBinaryTree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    max_num = 0
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        if not root:
            return 0
        self.find(root, 0)
        return self.max_num

    def find(self, node: TreeNode, idx: int):
        if not node.left and not node.right:
            return idx
        else:
            left, right = 0, 0
            if node.left:
                left = self.find(node.left, idx+1)
            if node.right:
                right = self.find(node.right, idx+1)
            cur_length = left-idx + right-idx
            self.max_num = max(self.max_num, cur_length)
        return max(left, right)

File: src/easy/Q543_DiameterofBinaryTree.java
package easy;

public class Q543_DiameterofBinaryTree {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    class Solution {
        int max = 0;
        public int diameterOfBinaryTree(TreeNode root) {
            max = 0;
            int left = getLength(root.left);
            int right =  getLength(root.right);
            return Math.max(max, left + right);
        }
        public int getLength(TreeNode node){
            if(node == null) return 0;
            int left = getLength(node.left);
            int right =  getLength(node.right);

            max = Math.max(max, left + right);
            return Math.max(left, right) + 1;
        }
    }
}
